avalon/tools/lib.py

```diff
@@ -19,16 +20,19 @@ def resource(*path):
 
 @contextlib.contextmanager
 def application():
-    app = QtWidgets.QApplication.instance()
+    # Storing QApplication to prevent garbage collection.
+    # See https://github.com/getavalon/core/issues/514
+    self.app = QtWidgets.QApplication.instance()
 
-    if not app:
-        print("Starting new QApplication..")
-        app = QtWidgets.QApplication(sys.argv)
-        yield app
-        app.exec_()
-    else:
+    if self.app:
         print("Using existing QApplication..")
-        yield app
+        yield self.app
+    else:
+        print("Starting new QApplication..")
+        self.app = QtWidgets.QApplication(sys.argv)
+        yield self.app
+
+    self.app.exec_()
```

Comment:
@tokejepsen Please, **don't always** run `self.app.exec_()`. This screws up the Blender integration. I also think it's the responsibility of the _creator_ of the `QApplication` to use `exec_()` if needed and shouldn't be called (again) for an existing `QApplication` instance.

Reply:
This is one of the issues that I'm seeing with the Photoshop integration. If we dont have that `exec_()` then I get an unresponsive window on the second launch.
Again not sure what the issue is with the unresponsive window.

What happens in Blender with always running `exec_()`?

Reply:
> This is one of the issues that I'm seeing with the Photoshop integration. If we dont have that exec_() then I get an unresponsive window on the second launch.
> Again not sure what the issue is with the unresponsive window.

The unresponsive window should be QApplication stopping with executing. This happens by default when the [last Qt window closes as described at the top of this comment](https://github.com/getavalon/core/pull/520#issuecomment-579528824).

---

@jasperges would love to understand too why `._exec()` would fail in Blender. From what I understood so far from Toke/David it will just return -1 in that case and disregard it.

Reply:
`.exec_()` doesn't fail, but it makes Blender unresponsive. After you quit all Qt Windows, Blenders GUI is responsive again.

Reply:
The problem that @jasperges describes should also be happening in C4D. In both those applications you should never trigger `QApplication.exec_()`.

Should we maybe push those behind an environment variable for those applications that should never trigger it? `AVALON_QTAPPLICATION_EXEC = False`

Potentially adding it in the host's `install()` function?

Reply:
> Should we maybe push those behind an environment variable for those applications that should never trigger it? AVALON_QTAPPLICATION_EXEC = False

Sounds good to me.

Reply:
Fine for me. One question though: why should the previous behaviour need to be changed at all?

If a new `QApplication` is spawned, run `QApplication.exec_()`. If an existing `QApplication` is already running, you might expect that it's event loop is also already running, right?

So apart from it causing issues with running Blender and Cinema4D (and possibly others), I don't understand why `QApplication.exec_()` should always be triggered.

Reply:
@jasperges see https://github.com/getavalon/core/pull/523#discussion_r375137717